src/formats/SMI.cpp

```diff
@@ -0,0 +1,705 @@
+/// Also not gaurenteed to find the SSSR, but it will find all rings
+/// needed for a given structure to be writen as SMILES
+/// \param adj_list The adjacentcy list graph
+/// \param hit_atoms Atoms already encountered
+/// \param ring_bonds Bonds already known to ring forming bonds
+/// \param ring_atoms Map from an atom to the number of rings it forms
+/// \param current_atom Atom to be checked in this iteration
+/// \param previous_atom Atom which was bound to the current atom
+static void find_rings_helper(
+    const std::vector<std::vector<size_t>>& adj_list,
+    std::vector<bool>& hit_atoms, std::set<Bond>& ring_bonds,
+    std::map<size_t, size_t>& ring_atoms,
+    size_t current_atom, size_t previous_atom) {
+
+    auto& current_atom_bonds = adj_list[current_atom];
+    hit_atoms[current_atom] = true;
+
+    for (auto neighbor : current_atom_bonds) {
+        // prevent back tracking
+        if (neighbor == previous_atom) {
+            continue;
+        }
+
+        // We've seen this neighbor before! Ring found
+        if (hit_atoms[neighbor]) {
+            // Don't process the same ring connection twice
+            if (ring_bonds.count(Bond(neighbor, current_atom))) {
+                continue;
+            }
+            ring_bonds.insert(Bond(neighbor, current_atom));
+
+            auto neigh_ring_iter = ring_atoms.find(neighbor);
+            if (neigh_ring_iter == ring_atoms.end()) {
+                neigh_ring_iter = ring_atoms.insert({neighbor, 0}).first;
+            }
+            ++neigh_ring_iter->second;
+
+            continue; // No need to see this atom again
+        }
+
+        // Continue the search
+        find_rings_helper(adj_list, hit_atoms, ring_bonds,
+                          ring_atoms, neighbor, current_atom);
+    }
+}
+
+static std::map<size_t, size_t> find_rings(
+    const std::vector<std::vector<size_t>>& adj_list) {
+
+    std::map<size_t, size_t> ring_atoms;
+    std::vector<bool> hit_atoms(adj_list.size(), false);
+    std::set<Bond> ring_bonds;
+    
+    while (!all(hit_atoms)) {
+        auto not_hit = std::find(hit_atoms.begin(), hit_atoms.end(), false);
+        auto current_atom = static_cast<size_t>(std::distance(hit_atoms.begin(), not_hit));
+        auto& current_atom_bonds = adj_list[current_atom];
+        hit_atoms[current_atom] = true;
+        if (current_atom_bonds.empty()) {
+            continue;
+        }
+
+        find_rings_helper(adj_list, hit_atoms, ring_bonds, ring_atoms,
+                          current_atom_bonds[0], current_atom
+        );
+    }
+
+    return ring_atoms;
+}
+
+static void write_atom_smiles(std::unique_ptr<TextFile>& file_, const Atom& atom) {
+    bool needs_brackets = false;
+    auto type = atom.type();
+
+    // The mass must be an integer is the only check we need as all atoms in the
+    // periodic table do have non-integer masses. Therefore, if the the mass is
+    // not an integer, then we know the user has set an isotope
```

Comment:
I am not sure I understand this comment. Pure atoms have integer mass, as well as isotopes. But the masses in `periodic_table.cpp` are not integers, since they are the average value with the natural isotopes occurrences.

Reply:
Sorry, the comment should have ended with 'Therefore, if the mass **is** an integer,  then we know the user has set an isotope.'